### PR TITLE
mgr/cephadm: fixing upgrade candidate versions listing

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/53527

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/tests/test_upgrade.py [hunk 1]
```diff
@@ -164,3 +164,48 @@ def test_enough_mds_for_ok_to_stop(get, get_daemons_by_service, cephadm_module:
...
+                                     'v17.1.0',
+                                     'v16.2.7',
+                                     'v16.2.6',
+                                     'v16.2.5',
+                                     'v16.1.4',
+                                     'v16.1.3',
+                                     'v15.2.0',
+                                 ],
+                                 (16, 1, '16.1.0'),
+                                 ['17.1.0', '16.2.7', '16.2.6', '16.2.5']
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/tests/test_upgrade.py:183]: I think we'd want to see 16.1.3 and 16.1.4 here. This implies somebody on 16.2.0 (initial pacific release) would not see 16.2.7 (most recent stable pacific release) which is not good. Just generally same major and minor version but higher z stream should definitely be present. Honestly, even if it's a lower z stream, if the major and minor version match we should probably show it.

Comment (rkachach) [reply src/pybind/mgr/cephadm/tests/test_upgrade.py]: Updated the code to also show all the releases with the same major/minor version (right now we don't check z stream... so we will show minor z streams if available).

**Review by adk3798 — APPROVED** [review]

File: src/pybind/mgr/cephadm/upgrade.py [hunk 2]
```diff
@@ -218,7 +224,10 @@ def upgrade_ls(self, image: Optional[str], tags: bool) -> Dict:
                     continue
                 if '-' in v[2]:
                     continue
-                versions.append('.'.join(v))
+                v_major = int(v[0])
+                v_minor = int(v[1])
+                if v_major > current_major or (v_major == current_major and v_minor >= current_minor):
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/upgrade.py:229]: https://pulpito.ceph.com/adking-2022-04-19_04:07:58-orch:cephadm-wip-adk-testing-2022-04-18-2113-distro-basic-smithi/6796317/
```
2022-04-19T05:29:18.936 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-04-19T05:29:18.937 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_ca2dd75f248daf78c6ea1f7118a600f23cfefd9c/qa/tasks/cephadm.py", line 1151, in shell
    extra_cephadm_args=args)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_ca2dd75f248daf78c6ea1f7118a600f23cfefd9c/qa/tasks/cephadm.py", line 49, in _shell
    **kwargs
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/orchestra/remote.py", line 509, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/orchestra/run.py", line 455, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/orchestra/run.py", line 161, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_b21f9558629bbbb865321b6e62e8b2c23e4d3291/teuthology/orchestra/run.py", line 183, in _raise_for_status
    node=self.hostname, label=self.label
teuthology.exceptions.CommandFailedError: Command failed on smithi059 with status 1: "sudo /home/ubuntu/cephtest/cephadm --image docker.io/ceph/ceph:v15.2.0 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 260b5c48-bf9f-11ec-8c38-001a4aab830c -- bash -c 'ceph orch upgrade ls --image quay.io/ceph/ceph | grep 16.2.0'"
```

From 
https://github.com/ceph/ceph/blob/master/qa/suites/orch/cephadm/upgrade/5-upgrade-ls.yaml
the test is expecting to see 16.2.0 and 16.2.2 in the upgrade ls output. The test needs to be modified along with this change so it still passes.

**Review by adk3798 — CHANGES_REQUESTED** [review]

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
@rkachach it looks like while you were making adjustments the signature on your commit was lost. Can you add that back?

**Comment by rkachach** [comment]
> @rkachach it looks like while you were making adjustments the signature on your commit was lost. Can you add that back?

yeah, actually I'm having some issues with vscode/squash these days as it's not storing correctly the resulting message. I fixed the signature manually.

**Comment by adk3798** [comment]
> > @rkachach it looks like while you were making adjustments the signature on your commit was lost. Can you add that back?
> 
> yeah, actually I'm having some issues with vscode/squash these days as it's not storing correctly the resulting message. I fixed the signature manually.

signed-ff-by is still failing. You have to surround your email address by <> in the commit message. So,

`Signed-off-by: Redouane Kachach rkachach@redhat.com`   ---->    `Signed-off-by: Redouane Kachach <rkachach@redhat.com>`

**Comment by adk3798** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test make check

**Pull request merged**